File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) 
    {
        if(matrix.length == 1) {
            
        }
        int rows = matrix.length;
        int cols = matrix[0].length;
        if (rows == 1) {
            return binarysearch(matrix, 0, 0, cols - 1, target);
        }
        int rStart = 0;
        int rEnd = rows - 1;
        int cMid = cols / 2;
        while (rStart < (rEnd - 1)) {
            int mid = rStart + (rEnd - rStart) / 2;
            if (matrix[mid][cMid] == target) {
                return true;
            }
            if (matrix[mid][cMid] < target) {
                rStart = mid;
            } else {
                rEnd = mid;
            }

        }
        if (matrix[rStart][cMid] == target) {
            return true;
        }
        if (matrix[rStart + 1][cMid] == target) {
            return true;
        }
        //search in 1st half
        if(cMid -1 >= 0 && target <= matrix[rStart][cMid - 1]){
            return binarysearch(matrix,rStart,0,cMid - 1,target);
        }
        //search in 2nd half
        if(cMid + 1 < matrix[0].length && target >= matrix[rStart][cMid + 1] && target <= matrix[rStart][cols - 1]){
            return binarysearch(matrix,rStart,cMid + 1,cols - 1,target);
        }
        //search in 3rd half
        if(cMid - 1 >= 0 && target <= matrix[rStart + 1][cMid - 1]){
            return binarysearch(matrix,rStart + 1,0,cMid - 1,target);
        }
        //search in 4th half
        else if(cMid + 1 < matrix[0].length) {
            return binarysearch(matrix,rStart + 1,cMid + 1,cols - 1,target);
        }
        else 
            return false;
    
    }
    
    static boolean binarysearch(int[][] matrix, int row, int cStart, int cEnd, int target) {
        while (cStart <= cEnd) {
            int mid = cStart + (cEnd - cStart) / 2;
            if (matrix[row][mid] == target){
                return true;
            }
            if (matrix[row][mid] < target){
                cStart = mid + 1;
            }
            else{
                cEnd = mid - 1;
            }
        }
        return false;
    }
}

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        

        rowStart = 0
        rowEnd = len(matrix) - 1
        colStart = 0
        colEnd = len(matrix[0]) - 1

        while rowStart <= rowEnd:
            mid = rowStart + (rowEnd - rowStart) // 2

            if matrix[mid][0] == target:
                return True
                
            if matrix[mid][0] > target:
                rowEnd = mid - 1
            else:
                if matrix[mid][colEnd] >= target:
                    while colStart <= colEnd:
                        x = colStart + (colEnd - colStart) // 2

                        if matrix[mid][x] == target:
                            return True
                        if matrix[mid][x] > target:
                            colEnd = x - 1
                        else:
                            colStart = x + 1
                    return False
                rowStart = mid + 1
        
        return False
        
